### ISSUE TITLE
Better diagnostic message for compilation failures while building rustdoc

### ISSUE THREAD

**Issue by VorpalBlade** [obi1kenobi/cargo-semver-checks#654]
### Steps to reproduce the bug with the above code

```bash
git clone https://github.com/archlinux/alpm.rs.git
cd alpm.rs
cargo semver-checks
```

Note you will need system libraries for libalpm installed to use this. libalpm is the library for the package mananger on Arch Linux. So it may be easiest to recreate in an arch linux docker/podman container for you (`docker.io/library/archlinux`). If so you will likely need to install build tooling:

```console
$ podman run -ti --rm docker.io/library/archlinux /bin/bash
$ # In container:
$ pacman -Syu --noconfirm base-devel rustup pkgconf git cmake
$ rustup install stable
$ cargo install cargo-semver-checks
$ # Same steps as above
```

### Actual Behaviour

```console
$ cargo semver-checks
     Parsing alpm v2.2.3 (current)
     Parsing alpm-sys v2.1.3 (current)
     Parsing alpm-utils v2.0.1 (current)
Error: running cargo-doc failed on /home/user/src/alpm/target/semver-checks/local-alpm-2_2_3/Cargo.toml:
   Compiling regex-syntax v0.8.2
   Compiling bindgen v0.66.1
   Compiling peeking_take_while v0.1.2
    Checking bitflags v1.3.2
   Compiling regex-automata v0.4.5
   Compiling regex v1.10.3
   Compiling alpm-sys v2.1.3 (/home/user/src/alpm/alpm-sys)
error: couldn't read /home/user/src/alpm/target/semver-checks/local-alpm-2_2_3/target/semver-checks/target/debug/build/alpm-sys-f51410450d14fd87/out/ffi_generated.rs: No such file or directory (os error 2)
 --> /home/user/src/alpm/alpm-sys/src/ffi_generated.rs:1:1
  |
1 | include!(concat!(env!("OUT_DIR"), "/ffi_generated.rs"));
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the macro `include` (in Nightly builds, run with -Z macro-backtrace for more info)

error: could not compile `alpm-sys` (lib) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
error: could not compile `alpm-sys` (lib) due to 1 previous error

```

### Expected Behaviour

Since the workspace passes `cargo build` and `cargo doc` I expect it to also work with `semver-checks`

#### Software version

cargo-semver-checks 0.28.0

#### Operating system

Linux 6.7.4-arch1-1

#### Command-line

```bash
/home/arvid/.cargo/bin/cargo-semver-checks semver-checks --bugreport 
```

#### cargo version

```
> cargo -V 
cargo 1.76.0 (c84b36747 2024-01-18)
```

#### Compile time information

- Profile: release
- Target triple: x86_64-unknown-linux-gnu
- Family: unix
- OS: linux
- Architecture: x86_64
- Pointer width: 64
- Endian: little
- CPU features: fxsr,sse,sse2
- Host: x86_64-unknown-linux-gnu

### Build Configuration

```console
$ cat ~/.cargo/config.toml              
[net]
git-fetch-with-cli = true

[build]
rustc-wrapper = "/usr/bin/sccache"

[target.x86_64-unknown-linux-gnu]
linker = "/usr/bin/clang"
rustflags = ["-C", "link-arg=--ld-path=/usr/bin/mold"]
```

**Comment by tonowak** [comment]
Looking into it :eyes: 

`cargo-semver-checks` seems to run the command `rustdoc  /tmp/alpm.rs/alpm-sys/src/lib.rs  --cfg 'feature="generate"'` (I've omitted other options for clarity), which fails with the 'error: environment variable `OUT_DIR` not defined at compile time' error.

**Comment by tonowak** [comment]
Running simply `cargo semver-checks --default-features` seems to work.

**Comment by VorpalBlade** [comment]
Oh, thanks. this is the first time I used cargo semver-checks, so sorry about not knowing how to use it. I wanted to tell if my PR would be a breaking change or not.

**Comment by VorpalBlade** [comment]
It does indeed seem like it fails when running `cargo build --all-features` as well.

**Comment by VorpalBlade** [comment]
Closing as it seems to be a alpm.rs bug not a semver check bug

**Comment by tonowak** [comment]
> so sorry about not knowing how to use it

I think `cargo-semver-checks` could do a better job with saying what went wrong :)

@obi1kenobi it might be not obvious that `cargo-semver-checks` runs the crates on more features than just the default ones. I wonder, when it checks a crate, could it also print the features it is testing? Usually the list of features is small, probably we can print them all in a single line. I currently don't have time to implement a prototype, unfortunately.

**Comment by tonowak** [comment]
Prefixing the output from the error with an information that `cargo-sermver-checks` encountered an external error might also be useful for debugging.

**Comment by obi1kenobi** [comment]
Thanks for triaging this and getting to the bottom of it!

Probably suffixing with error info would work even better, because the external error printout might be huge and people might not see the prefix.

Perhaps a suffix like:
```
error: failed to build rustdoc for crate alpm v2.2.3

note: this is usually due to a compilation error in the crate
note: running the following command on the crate may reproduce the error:
    cargo build --features a,b,c,d
```

What do you think?

**Comment by tonowak** [comment]
> Probably suffixing with error info would work even better

Maybe both? For me, it was not clear where the output of cargo-semver-checks stopped, and where the error completely unrelated to `cargo-semver-checks` started.

> Is this clear?

Yes.

> note: running the following command on the crate may reproduce the error:

Why `may`? It was the exact command used. `Should` is probably closer to the truth.

**Comment by obi1kenobi** [comment]
New suggestion attempting to address all the feedback:
```
error: running cargo-doc failed on /home/user/src/alpm/target/semver-checks/local-alpm-2_2_3/Cargo.toml with output:
***
   Compiling regex-syntax v0.8.2
   Compiling bindgen v0.66.1
   Compiling peeking_take_while v0.1.2
    Checking bitflags v1.3.2
   Compiling regex-automata v0.4.5
   Compiling regex v1.10.3
   Compiling alpm-sys v2.1.3 (/home/user/src/alpm/alpm-sys)
error: couldn't read /home/user/src/alpm/target/semver-checks/local-alpm-2_2_3/target/semver-checks/target/debug/build/alpm-sys-f51410450d14fd87/out/ffi_generated.rs: No such file or directory (os error 2)
 --> /home/user/src/alpm/alpm-sys/src/ffi_generated.rs:1:1
  |
1 | include!(concat!(env!("OUT_DIR"), "/ffi_generated.rs"));
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the macro `include` (in Nightly builds, run with -Z macro-backtrace for more info)

error: could not compile `alpm-sys` (lib) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
error: could not compile `alpm-sys` (lib) due to 1 previous error
***

error: failed to build rustdoc for crate alpm v2.2.3

note: this is usually due to a compilation error in the crate
note: running the following command on the crate should reproduce the error:
    cargo build --features a,b,c,d
```

**Comment by tonowak** [comment]
Looks good!

**Comment by obi1kenobi** [comment]
After #656 merges and if one of you has a second, I'd love it if you could try pulling `main` and checking if the new error message appears as expected and is now satisfactory.

**Comment by obi1kenobi** [comment]
The final output should look like the following:
```
error: running cargo-doc failed on /home/user/src/alpm/target/semver-checks/local-alpm-2_2_3/Cargo.toml with output:
-----
   Compiling regex-syntax v0.8.2
   Compiling bindgen v0.66.1
   Compiling peeking_take_while v0.1.2
    Checking bitflags v1.3.2
   Compiling regex-automata v0.4.5
   Compiling regex v1.10.3
   Compiling alpm-sys v2.1.3 (/home/user/src/alpm/alpm-sys)
error: couldn't read /home/user/src/alpm/target/semver-checks/local-alpm-2_2_3/target/semver-checks/target/debug/build/alpm-sys-f51410450d14fd87/out/ffi_generated.rs: No such file or directory (os error 2)
 --> /home/user/src/alpm/alpm-sys/src/ffi_generated.rs:1:1
  |
1 | include!(concat!(env!("OUT_DIR"), "/ffi_generated.rs"));
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the macro `include` (in Nightly builds, run with -Z macro-backtrace for more info)

error: could not compile `alpm-sys` (lib) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
error: could not compile `alpm-sys` (lib) due to 1 previous error
-----

error: failed to build rustdoc for crate alpm v2.2.3

note: this is usually due to a compilation error in the crate,
      and is unlikely to be a bug in cargo-semver-checks
note: running the following command on the crate should reproduce the error:
      cargo build --no-default-features --features a,b,c,d
```

**Comment by tonowak** [comment]
Now I am confused. Should I locally run `cargo run -- semver-checks --baseline-root=../alpm.rs/`? But it fails with
```
    Finished dev [unoptimized + debuginfo] target(s) in 0.14s
     Running `target/debug/cargo-semver-checks semver-checks --baseline-root=../alpm.rs/`
     Parsing cargo-semver-checks v0.28.0 (current)
      Parsed [   3.080s] (current)
error: possibly due to errors: [
  failed to parse ../alpm.rs/Cargo.toml: no `package` table,
]

Caused by:
    package `cargo-semver-checks` not found in ../alpm.rs/
```

**Comment by tonowak** [comment]
OK, I got it. Here's the output:
```
➜  cargo-semver-checks git:(main) cargo run -- semver-checks --manifest-path=../alpm.rs/Cargo.toml
    Finished dev [unoptimized + debuginfo] target(s) in 0.15s
     Running `target/debug/cargo-semver-checks semver-checks --manifest-path=../alpm.rs/Cargo.toml`
     Parsing alpm v2.2.3 (current)
error: running cargo-doc on crate alpm failed with output:
-----
   Compiling alpm-sys v2.1.3 (/tmp/alpm.rs/alpm-sys)
error: couldn't read /tmp/alpm.rs/target/semver-checks/local-alpm-2_2_3/target/semver-checks/target/debug/build/alpm-sys-0c58d195ab46ab64/out/ffi_generated.rs: No such file or directory (os error 2)
 --> /tmp/alpm.rs/alpm-sys/src/ffi_generated.rs:1:1
  |
1 | include!(concat!(env!("OUT_DIR"), "/ffi_generated.rs"));
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the macro `include` (in Nightly builds, run with -Z macro-backtrace for more info)

error: could not compile `alpm-sys` (lib) due to previous error
warning: build failed, waiting for other jobs to finish...
error: could not compile `alpm-sys` (lib) due to previous error

-----

error: failed to build rustdoc for crate alpm v2.2.3
note: this is usually due to a compilation error in the crate,
      and is unlikely to be a bug in cargo-semver-checks
note: running the following command on the crate should reproduce the error:
      cargo build --no-default-features --features checkver,default,docs-rs,generate,git,libarchive,libarchive3-sys,mtree,static
     Parsing alpm-sys v2.1.3 (current)
error: running cargo-doc on crate alpm-sys failed with output:
-----
 Documenting alpm-sys v2.1.3 (/tmp/alpm.rs/alpm-sys)
error: couldn't read /tmp/alpm.rs/target/semver-checks/local-alpm_sys-2_1_3/target/semver-checks/target/debug/build/alpm-sys-0c58d195ab46ab64/out/ffi_generated.rs: No such file or directory (os error 2)
 --> /tmp/alpm.rs/alpm-sys/src/ffi_generated.rs:1:1
  |
1 | include!(concat!(env!("OUT_DIR"), "/ffi_generated.rs"));
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the macro `include` (in Nightly builds, run with -Z macro-backtrace for more info)

error: could not document `alpm-sys`

Caused by:
  process didn't exit successfully: `/home/tonowak/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/bin/rustdoc --crate-type lib --crate-name alpm_sys /tmp/alpm.rs/alpm-sys/src/lib.rs -o /tmp/alpm.rs/target/semver-checks/local-alpm_sys-2_1_3/target/semver-checks/target/doc --cfg 'feature="bindgen"' --cfg 'feature="docs-rs"' --cfg 'feature="generate"' --cfg 'feature="git"' --cfg 'feature="static"' --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat -C metadata=c53731c23b962bdc -L dependency=/tmp/alpm.rs/target/semver-checks/local-alpm_sys-2_1_3/target/semver-checks/target/debug/deps -Z unstable-options --document-private-items --document-hidden-items --output-format=json --cap-lints=allow --crate-version 2.1.3` (exit status: 1)

-----

error: failed to build rustdoc for crate alpm-sys v2.1.3
note: this is usually due to a compilation error in the crate,
      and is unlikely to be a bug in cargo-semver-checks
note: running the following command on the crate should reproduce the error:
      cargo build --no-default-features --features docs-rs,generate,git,static
     Parsing alpm-utils v2.0.1 (current)
error: running cargo-doc on crate alpm-utils failed with output:
-----
   Compiling alpm-sys v2.1.3 (/tmp/alpm.rs/alpm-sys)
error: couldn't read /tmp/alpm.rs/target/semver-checks/local-alpm_utils-2_0_1/target/semver-checks/target/debug/build/alpm-sys-0c58d195ab46ab64/out/ffi_generated.rs: No such file or directory (os error 2)
 --> /tmp/alpm.rs/alpm-sys/src/ffi_generated.rs:1:1
  |
1 | include!(concat!(env!("OUT_DIR"), "/ffi_generated.rs"));
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the macro `include` (in Nightly builds, run with -Z macro-backtrace for more info)

error: could not compile `alpm-sys` (lib) due to previous error
warning: build failed, waiting for other jobs to finish...
error: could not compile `alpm-sys` (lib) due to previous error

-----

error: failed to build rustdoc for crate alpm-utils v2.0.1
note: this is usually due to a compilation error in the crate,
      and is unlikely to be a bug in cargo-semver-checks
note: running the following command on the crate should reproduce the error:
      cargo build --no-default-features --features alpm,conf,default,docs-rs,generate,git,pacmanconf,static
error: aborting due to failure to build rustdoc for crate alpm v2.2.3
```

**Comment by obi1kenobi** [comment]
Nice, thanks! I'm going to throw an extra newline between the "reproduce the error" command and the next line, just to make it a bit easier to read when output continues.
